text/0000-cargo-features2.md

```diff
@@ -0,0 +1,519 @@
+
+When a package appears multiple times in the build graph, and one of those
+instances is a target-specific dependency, then the features of the
+target-specific dependency are only enabled if the target is currently being
+built. For example:
+
+```toml
+[dependency.common]
+version = "1.0"
+features = ["f1"]
+
+[target.'cfg(windows)'.dependencies.common]
+version = "1.0"
+features = ["f2"]
+```
+
+When building this example for a non-Windows platform, the `f2` feature will
+*not* be enabled.
+
+### dev-dependencies
+
+When a package is shared as a normal dependency and a dev-dependency, the
+dev-dependency features are only enabled if the current build is including
+dev-dependencies. For example:
+
+```toml
+[dependencies]
+serde = {version = "1.0", default-features = false}
+
+[dev-dependencies]
+serde = {version = "1.0", features = ["std"]}
+```
+
+In this situation, a normal `cargo build` will build `serde` without any
+features. When built with `cargo test`, Cargo will build `serde` with its
+default features plus the "std" feature.
+
+Note that this is a global decision. So a command like `cargo build
+--all-targets` will include examples and tests, and thus features from
+dev-dependencies will be enabled.
+
+### Host dependencies
+
+When a package is shared as a normal dependency and a build-dependency or
+proc-macro, the features for the normal dependency are kept independent of the
+build-dependency or proc-macro. For example:
+
+```toml
+[dependencies]
+log = "0.4"
+
+[build-dependencies]
+log = {version = "0.4", features=['std']}
+```
+
+In this situation, the `log` package will be built with the default features
+for the normal dependencies. As a build-dependency, it will have the `std`
+feature enabled. This means that `log` will be built twice, once without `std`
+and once with `std`.
+
+Note that a dependency shared between a build-dependency and proc-macro are
+still unified. This is intended to help reduce build times, and is expected to
+be unlikely to cause problems that feature unification usually cause because
+they are both being built for the host compiler, and are only used at build
+time.
+
+## Resolver opt-in
+
+Testing has been performed on various projects. Some were found to fail to
+compile with the new resolver. This is because some dependencies are written
+to assume that features are enabled from another part of the graph. Because
+the new resolver results in a backwards-incompatible change in resolver
+behavior, the user must opt-in to use the new resolver. This can be done with
+the `resolver` field in `Cargo.toml`:
+
+```toml
+[package]
+name = "my-package"
+version = "1.0.0"
+resolver = "2"
+```
+
+Setting the resolver to `"2"` switches Cargo to use the new feature resolver.
+It also enables backwards-incompatible behavior detailed in [New command-line
+behavior](#new-command-line-behavior). `"2"` is the only valid option, if
+`resolver` is not specified then the old behavior is used.
```

Comment:
Is there any reason why `resolver = "1"` is not something we are going to accept, even if just for reduced surprise factor? (As a prior art `--edition` accepts both 2018 and 2015 as a value)

Reply:
There isn't a strong reason, just a personal preference of not wanting to see `resolver = "1"` being used, but it can be easily added.

Reply:
I think we should add this, so that if we want to change the default for `resolver` at a later date (maybe a new rust edition etc..)

Reply:
@pksunkara That assumes we want to support using a new edition but opting for the old resolver, which I don't know that we want to support. We could minimize the matrix of simultaneously supported features by not supporting the old resolver on a new edition that implies the new resolver.